### ISSUE TITLE
Calculate extinct flag

### ISSUE THREAD

**Issue by mdoering** [CatalogueOfLife/backend#870]
It seems that extinct=false actually filters out all higher taxa (i.e. not marked explicitly with extinct=false)
Example nameusage search

https://api.dev.catalogue.life/dataset/1008/nameusage/search?facet=extinct&extinct=false
https://api.dev.catalogue.life/dataset/1008/nameusage/search?facet=extinct&extinct=true
https://api.dev.catalogue.life/dataset/1008/nameusage/search?facet=extinct&extinct=

**Comment by mdoering** [comment]
Hm, I don't think it is that wrong.
There are 13791 hits with extinct=false, i.e. 13791 usages with a clear flag saying the taxon is living.
Then there are 0 usages with extinct=true, so none.
And 31569 alltogether including nulls.

@thomasstjerne it might make sense to offer only 2 states for the user, extinct=true or null.
There are not many records where people actually positively assert the taxon is living. 
It is a flag for extinct only after all. 

@yroskov @gdower @dhobern guess we could make it a primitive boolean with just 2 states yes/no and simply default to not extinct in case the flag is not explicitly set? Is it intuitive or useful for a user to deal with nulls, i.e. unknown?

**Comment by yroskov** [comment]
> guess we could make it a primitive boolean with just 2 states yes/no and simply default to not extinct in case the flag is not explicitly set? Is it intuitive or useful for a user to deal with nulls, i.e. unknown?

Only extinct species and higher taxa with all extinct children species should get "Extinct" comment on species page and dagger in the Tree, etc.. Neither extant, nor taxa with status "unknown" should get any comments in public view.

**Comment by mdoering** [comment]
Agree. So we should really derive the extinct flag based on the species data then and propagate extinct upwards if all included taxa are extinct. We could do that easily on release time.

Does it make sense in the API to differ between unknown and extant or would it not be better to have extinct/extant only?

**Comment by mdoering** [comment]
As for propagation of the extinct flag upwards from species I do wonder if that is a good idea in all cases?
Could it not be the other way around too, that a dataset declares Trilobita as extinct and all included taxa should then be marked as extinct even if they have nulls? We surely want this as an editorial decision. But maybe also driven from the original sources?

**Comment by mdoering** [comment]
Basically I see 3 options:
 1) we simply keep extinct flags as they were given. This will likely lead to inconsistencies, e.g. Trilobita being marked as extinct but some included taxa not.
 2) we trust species data and propagate the extinct flag upwards if all included taxa are extinct. This will likely cause Trilobita to not be extinct if a single species is missing the extinct flag. This can be fixed through editorial decisions though.
 3) we trust higher taxa more than species and inherit the extinct flag to included taxa if it was set. That allows Trilobita to be marked as extinct and all included children also, no matter if they have a flag given in the sources. Useful if we merge names from different sectors and sources that might not publish the extinct flag at all to COL.

All 3 options have issues.

**Comment by dhobern** [comment]
The last of these three models is part of the reason why IRMNG is what it is. It makes data curation possible in ways that are very hard with the first and second models.

**Comment by mdoering** [comment]
Yes. But we do have the options of leaving a manual editorial decision for a higher taxon that pushes the flag down just like number 3 would. And that works with any of the 3 models which are modes for the default, automated behavior without any intervening, manual decision. In that light maybe number 2 is the best for us if it is combined with editorial top down corrections?

**Comment by yroskov** [comment]
> As for propagation of the extinct flag upwards from species I do wonder if that is a good idea in all cases?

In CoL procedures, extinct status of the parent was automatically defined by complete set of extinct children species. This protocol worked well in old protocols and software. We never had problems with this, and I would like to follow the same protocols: we are checking that extinct flag is present on species level, software populate it appropriately to higher taxa.

> Could it not be the other way around too, that a dataset declares Trilobita as extinct and all included taxa should then be marked as extinct even if they have nulls? We surely want this as an editorial decision. But maybe also driven from the original sources? 

Flag for extinct species is not an editorial decision. It should be captured correctly from the source GSD at the process of crawling/import. I would like to be able to do extinct flag correction via Complex Decision tool for only occasional cases (i.e. to fix mistakes in 1-5 species, not more).

**Comment by dhobern** [comment]
I think this can definitely be an editorial decision. We must distinguish between three possible assertions:

- Every species in group X is extinct (a real world fact that represents what most people want to know, but that cannot directly be proven)
- The best scientific opinion is that every species in group X is extinct (for many groups, something that is easily determined)
- Experts contributing to COL have indicated that every listed species in group X is extinct (error-prone since it is easy for a single mistaken flag to break the whole inference, and in any case it is likely that almost every GSD is missing one or more named species, so the inference is being made on a partial set)

We should be using bidirectional validation to confirm consistency and then fix either the species-level records or the IRMNG-level assertions to match the reality. However, the safest approach, expecially for the higher-level taxa that fall within the COL higher taxonomy, will be at very least to track and report assertions at that level that the entire taxon is extinct. The same people clade-based decisions made in the upper classification may in fact determine the truth or falsity for some extinction assertions.

**Comment by mdoering** [comment]
As it stands the assembly process does not apply any automated corrections, inheritance or propagation.

**Comment by mdoering** [comment]
looks like we had an open issue already on this: https://github.com/CatalogueOfLife/backend/issues/289

**Comment by mdoering** [comment]
My proposal would be:
 - by default propagate the extinct flag from terminal taxa upwards as long as all children are marked as extinct
 - allow manual decisions on any rank that apply to all descendants. This way we can mark a larger group like trilobites if needed without the need to manually fix various species.

And to do sth very similar for environment assertions.

**Comment by mdoering** [comment]
duplicate of https://github.com/CatalogueOfLife/backend/issues/289
